src/Cli/Microsoft.DotNet.Cli.Utils/PathUtility.cs

```
@@ -416,4 +416,60 @@ public static string GetDirectorySeparatorChar()
 
         return null;
     }
+
+    public static void SafeRenameDirectory(string sourceDir, string destDir)
```

I haven't had time yet to review the implementation in detail, or to see how this is used, but, this looks like something difficult to get right. File systems can do weird stuff--hard links, mount points, etc.

Is this really the kind of thing where we want to automatically figure out what to do, rather than just complaining that the file system isn't in a state where we can do it automatically, and user needs to make the appropriate change to allow the tool to work?

---

I'm not opposed to failing more instead of trying to be smart. In this case, I see these options:
1. Simply call `Directory.Move` without any checks. That will fail with an exception in case we are trying to move a directory into itself. And you are left in half-migrated state. We don't want that.
2. Do some checks *up front* and fail early. So we still need some logic to detect that `Directory.Move` won't work. Since we already need to detect this situation anyway, why not do the automatic thing (moving to temp directory as a middle step)? I'm not sure, just curious what you think.
3. Have this `SafeRenameDirectory` utility. It seems it might be useful in other cases for the CLI - basically if you have a directory like `/a/b`, you should be able to "rename" it to `/a/b/c` as programmer and this utility allows that. For example, vscode file explorer can do this as well.